### PR TITLE
test(mcp): unit-test private parsers + proptest truncate & parse_embedding

### LINKED ISSUE #317: [super-qa] high: 1387-line tools module has no #[cfg(test)] block — private parsing helpers are untested at unit level
> ⚠️ Generated by `/super-qa` 2026-06-01. Severity & details from agent analysis; **not independently source-verified**. Review before action.

**1387-line tools module has no #[cfg(test)] block — private parsing helpers are untested at unit level**

- **ID:** `mcp/testing-inline-tools-no-unit-tests`
- **Severity:** high
- **Area:** mcp · **Type:** test
- **Location:** `memory-engine-mcp/src/tools/mod.rs`
- **Auto-fixable:** no

## Description
tools/mod.rs is the largest file in the crate (1387 lines) and contains a large set of private helper functions: parse_event_type (line 488), parse_fact_type (line 498), parse_search_mode (line 476), parse_replay_order (line 1120), parse_outcome (line 1255), parse_embedding (line 467), get_datetime (line 445), and require_f64_if_present (line 509). None has a dedicated #[cfg(test)] mod. Integration tests exercise these transitively but only over the success/known-error paths. Fine-grained boundary cases — e.g., parse_embedding with a JSON array containing non-numeric values, get_datetime with malformed-but-not-empty strings, require_f64_if_present distinguishing absent vs wrong-type (the bug it was written to catch) — are exercised only incidentally or not at all.

## Suggested approach
Add a `#[cfg(test)] mod tests` block at the bottom of tools/mod.rs. At a minimum: (1) one happy-path + one error path for each parse_* fn; (2) `require_f64_if_present` with `Value::String("1".into())` must return Err (that is the exact regression the fn guards); (3) `get_datetime` with an invalid ISO-8601 string must return Err; (4) `parse_embedding` with a JSON array containing a string element must return Err.

## References
memory-engine-mcp/src/tools/mod.rs:467-516, memory-engine-mcp/src/tools/mod.rs:476-496, memory-engine-mcp/src/tools/mod.rs:488-505, memory-engine-mcp/src/tools/mod.rs:1120-1128, memory-engine-mcp/src/tools/mod.rs:1255-1265

### LINKED ISSUE #471: [super-qa] medium: Strong property-based testing candidates have only example-based tests
> ⚠️ Generated by `/super-qa` 2026-06-01. Severity & details from agent analysis; **not independently source-verified**. Review before action.

**Strong property-based testing candidates have only example-based tests**

- **ID:** `mcp/testing-proptest-candidates`
- **Severity:** medium
- **Area:** mcp · **Type:** test
- **Location:** `memory-engine-mcp/src/depth.rs:27-37, memory-engine-mcp/src/tools/mod.rs:467-474`
- **Auto-fixable:** no

## Description
Two functions are strong proptest candidates: (1) depth::truncate (lines 27-37) — the fn must guarantee both output.len() <= max AND s.is_char_boundary(output.len()) for all UTF-8 inputs; the existing unit test at depth.rs:405-411 uses a single hand-crafted emoji string and cannot explore the full character boundary space. (2) parse_embedding (tools/mod.rs:467-474) — for any Vec<f32> serialised to JSON then parsed back, the round-trip must yield the exact same vector; currently only happy-path dispatch tests exercise it. proptest is already present in the root crate's dev-dependencies (Cargo.toml line 24) but the mcp crate does not declare it.

## Suggested approach
Add `proptest = "1"` to [dev-dependencies] in memory-engine-mcp/Cargo.toml. Then add proptest tests in depth.rs: `proptest! { #[test] fn truncate_always_valid_utf8(s in ".*", max in 0usize..500) { let t = truncate(&s, max); assert!(t.len() <= max); assert!(s.is_char_boundary(t.len())); } }`. Add a round-trip test for parse_embedding: `proptest! { #[test] fn embedding_roundtrip(vals in prop::collection::vec(-1.0f32..1.0, 1..128)) { let json = serde_json::to_value(&vals).unwrap(); let map = [("embedding", json)].into_iter().collect(); let result = parse_embedding(&map).unwrap().unwrap(); assert_eq!(result, vals); } }`

## References
memory-engine-mcp/src/depth.rs:27-37, memory-engine-mcp/src/depth.rs:405-411, memory-engine-mcp/src/tools/mod.rs:467-474, memory-engine/Cargo.toml

### PR DESCRIPTION
**Pull request by dutiona**
## Summary

Fills the test gaps in the MCP server's private parsers and adds the first property-based tests to the crate.

### #317 — unit tests for uncovered private parsers (`tools/mod.rs`)

The `#[cfg(test)] mod tests` block already covered `parse_fact_type`, `parse_consolidate_config`, and `default_dump*`. This adds happy- and error-path unit tests for the parsers that had none:

- `parse_event_type` — every known variant → `Ok`; an unknown string → `Err` (offending token preserved).
- `parse_outcome` — every known variant → `Ok`; an unknown string → `Err` (token preserved).
- `parse_embedding` — a correctly-sized array with a **non-numeric element** passes the length gate but fails `Vec<f32>` deserialization → `Err` (the wrong-length and non-array cases were already covered).
- `get_datetime` — absent → `Ok(None)`; valid ISO → round-trips; a **malformed-but-non-empty** ISO string → `Err`.
- `require_f64_if_present` — absent → `Ok(None)`; numeric → `Ok(Some)`; `Value::String("1")` present → `Err` (the exact silent-fallback regression this guard exists for).

### #471 — property-based tests with `proptest`

- **`depth::truncate`** (`depth.rs`): for any `s: String` and any `max in 0..500`, `truncate(&s, max)` satisfies `output.len() <= max` **and** `s.is_char_boundary(output.len())` — the invariant the byte back-off loop upholds (a non-boundary slice would panic).
- **`parse_embedding` round-trip** (`tools/mod.rs`): any fixed-dim (`D = 8`) `Vec<f32>` of finite values serializes to JSON, parses back through `parse_embedding(&map, D)`, and equals the input bit-for-bit. serde_json widens each f32 to f64 for the JSON text and narrows back; that round-trip is exact for every finite f32 (f64 represents all f32 losslessly), so strict equality is the correct contract. Non-finite values are excluded (serde_json cannot encode NaN/±inf as JSON numbers).

`proptest = "1"` added to `[dev-dependencies]`.

## Gate

- `cargo test -p memory-engine-mcp` — 180 passed (lib suite: 57 passed, including all new unit tests + both proptests).
- `cargo clippy -p memory-engine-mcp --all-targets` — clean (the 2 remaining warnings are the pre-existing `memory-engine` lib storage warnings, #843).
- `cargo fmt --check` — clean.
- `cargo build --workspace` — green.
- `cargo deny check` — advisories ok, bans ok, licenses ok, sources ok (proptest introduces no new advisory or license issue).

Closes #317
Closes #471
Part of #256

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: memory-engine/bin/mcp/src/depth.rs [hunk 1]
```diff
@@ -520,4 +520,23 @@ mod tests {
...
+
+    // --- Property-based tests (#471) ---
+
+    proptest::proptest! {
+        /// `truncate` must never exceed `max` bytes and must always cut on a valid
+        /// UTF-8 char boundary, for *any* input string and any cap in 0..500. This
+        /// covers the contract the byte-boundary back-off loop exists to uphold —
+        /// slicing on a non-boundary would panic at runtime.
+        #[test]
+        fn truncate_respects_cap_and_char_boundary(s in ".*", max in 0_usize..500) {
```
[the remark this answers is not in the document]
Comment (dutiona) [reply memory-engine/bin/mcp/src/depth.rs:534]: Strengthened in `2085722`: the proptest now asserts `s.starts_with(out)` (genuine prefix) plus maximality — if `s` fits, `out == s`; else the next char would overflow `max`. This kills the always-empty and off-by-one mutants the prior `len <= max` + `is_char_boundary` pair survived. Verified passing against the real `truncate` at 4096 cases.
